src/client.rs

```diff
@@ -769,8 +773,44 @@ impl<C: Config> Client<C> {
             }
             C2sPlayPacket::CommandSuggestionsRequest(_) => {}
             C2sPlayPacket::ClickContainerButton(_) => {}
-            C2sPlayPacket::ClickContainer(_) => {}
-            C2sPlayPacket::CloseContainerC2s(_) => {}
+            C2sPlayPacket::ClickContainer(p) => {
+                if p.slot_idx == -999 {
+                    // client is trying to drop the currently held stack
+                    let held = self.cursor_held_item.clone();
+                    self.cursor_held_item = Slot::Empty;
+                    match held {
+                        crate::protocol::Slot::Empty => {}
+                        crate::protocol::Slot::Present {
+                            item_id,
+                            item_count,
+                            nbt,
+                        } => self.events.push_back(ClientEvent::DropItemStack {
+                            item_id,
+                            item_count,
+                            nbt,
+                        }),
+                    }
+                } else {
+                    self.cursor_held_item = p.carried_item.clone();
+                    self.events.push_back(ClientEvent::ClickContainer {
+                        window_id: p.window_id,
+                        state_id: p.state_id,
+                        slot_id: p.slot_idx as u16,
+                        mode: p.mode,
+                        slot_changes: p
+                            .slots
+                            .into_iter()
+                            .map(|(slot_id, slot)| (slot_id as u16, slot))
```

Comment:
Some reasons to use `i16` for slot indices instead of `u16`
- Conversions from `slot_id` to `u16` can underflow because we're not checking that `slot_id >= 0` in this code. It wouldn't cause errors per se, but it might not be what people expect. In the future `inventory` module I plan on having a type alias like

  ```rust
  pub type SlotIdx = i16;
  ```
- Don't have to worry about underflows when doing math on slot indices as much.
- We can move `slots` directly from the packet to the event without allocating a new `Vec`.